File: packages/core/src/permission.test.ts
// Ported 1:1 from backend/internal/core/permission/permission_test.go.
import { describe, expect, it } from 'vitest';
import {
  type Actor,
  actorFromClaims,
  canManageAdmin,
  canReadAll,
  canReadDivision,
  canWrite,
  isLead,
  makeRole,
} from './permission.js';

const actor = (employeeId: string, role: Parameters<typeof makeRole>[0]): Actor => ({
  employeeId,
  role: makeRole(role),
});

describe('role matrix', () => {
  const staff = actor('S', { division: 'Creative', level: 'staff' });
  const lead = actor('L', { division: 'Creative', level: 'lead' });
  const od = actor('O', { od: true });
  const director = actor('D', { director: true });
  // Layered case: one employee is Creative Staff AND OD.
  const staffOD = actor('SO', { division: 'Creative', level: 'staff', od: true });

  const cases: Array<{
    name: string;
    a: Actor;
    canWrite: boolean;
    admin: boolean;
    readAll: boolean;
    lead: boolean;
    readOwnDiv: boolean;
    readOtherDiv: boolean;
  }> = [
    { name: 'staff', a: staff, canWrite: true, admin: false, readAll: false, lead: false, readOwnDiv: false, readOtherDiv: false },
    { name: 'lead', a: lead, canWrite: true, admin: false, readAll: false, lead: true, readOwnDiv: true, readOtherDiv: false },
    { name: 'od', a: od, canWrite: false, admin: false, readAll: true, lead: false, readOwnDiv: true, readOtherDiv: true },
    { name: 'director', a: director, canWrite: true, admin: true, readAll: true, lead: true, readOwnDiv: true, readOtherDiv: true },
    { name: 'staff+od', a: staffOD, canWrite: true, admin: false, readAll: true, lead: false, readOwnDiv: true, readOtherDiv: true },
  ];

  it.each(cases)('$name', (c) => {
    expect(canWrite(c.a)).toBe(c.canWrite);
    expect(canManageAdmin(c.a)).toBe(c.admin);
    expect(canReadAll(c.a)).toBe(c.readAll);
    expect(isLead(c.a, 'Creative')).toBe(c.lead);
    expect(canReadDivision(c.a, 'Creative')).toBe(c.readOwnDiv);
    expect(canReadDivision(c.a, 'Ads')).toBe(c.readOtherDiv);
  });
});

// The layered OD scope must never confer write; only the underlying division
// scope does. A pure OD (no division) cannot write.
describe('layered OD never writes from OD scope', () => {
  it('pure OD is read-only', () => {
    const pureOD = actor('O', { od: true });
    expect(canWrite(pureOD)).toBe(false);
  });

  it('Staff+OD writes from staff scope but is not a lead', () => {
    const staffOD = actor('SO', { division: 'Creative', level: 'staff', od: true });
    expect(canWrite(staffOD)).toBe(true);
    expect(isLead(staffOD, 'Creative')).toBe(false);
  });
});

// actorFromClaims must mirror employee_claims (SQL) and ResolveActor (Go): the
// same five app_metadata keys mapped identically. These cases pin that parity.
describe('actorFromClaims (JWT app_metadata → Actor)', () => {
  it('maps a mapped staff', () => {
    const a = actorFromClaims({ employee_id: 'EMP-1', division: 'Creative', level: 'staff', od: false, director: false });
    expect(a).toEqual({ employeeId: 'EMP-1', role: { division: 'Creative', level: 'staff', od: false, director: false } });
    expect(canWrite(a)).toBe(true);
    expect(isLead(a, 'Creative')).toBe(false);
  });

  it('maps a division lead', () => {
    const a = actorFromClaims({ employee_id: 'EMP-2', division: 'Ads', level: 'lead' });
    expect(isLead(a, 'Ads')).toBe(true);
    expect(isLead(a, 'Creative')).toBe(false);
  });

  it('maps a pure Director (no mapping) — director true, empty division/level', () => {
    const a = actorFromClaims({ employee_id: 'EMP-3', division: '', level: '', od: false, director: true });
    expect(a.role).toEqual({ division: '', level: '', od: false, director: true });
    expect(canManageAdmin(a)).toBe(true);
    expect(canReadAll(a)).toBe(true);
    expect(isLead(a, 'Whatever')).toBe(true); // director carries lead authority everywhere
  });

  it('maps a layered Staff+OD', () => {
    const a = actorFromClaims({ employee_id: 'EMP-4', division: 'Creative', level: 'staff', od: true });
    expect(canReadAll(a)).toBe(true);
    expect(canWrite(a)).toBe(true); // writes from the underlying division scope
  });

  it('coalesces missing division/level to empty (never elevated)', () => {
    const a = actorFromClaims({ employee_id: 'EMP-5' });
    expect(a.role).toEqual({ division: '', level: '', od: false, director: false });
  });

  it('rejects an unknown level to empty scope (no privilege inference)', () => {
    const a = actorFromClaims({ employee_id: 'EMP-6', level: 'superadmin' });
    expect(a.role.level).toBe('');
  });

  it('coerces non-boolean od/director to false', () => {
    const a = actorFromClaims({ employee_id: 'EMP-7', od: 'true', director: 1 });
    expect(a.role.od).toBe(false);
    expect(a.role.director).toBe(false);
  });

  it('throws when employee_id is missing or empty (unresolved actor)', () => {
    expect(() => actorFromClaims({})).toThrow(/employee_id missing/);
    expect(() => actorFromClaims({ employee_id: '' })).toThrow(/employee_id missing/);
    expect(() => actorFromClaims(null)).toThrow(/employee_id missing/);
    expect(() => actorFromClaims(undefined)).toThrow(/employee_id missing/);
  });
});

File: archive/backend-go/internal/core/permission/permission_test.go
package permission_test

import (
	"testing"

	"github.com/meagrup/agencyapp/backend/internal/core/permission"
)

func TestRoleMatrix(t *testing.T) {
	staff := permission.Actor{EmployeeID: "S", Role: permission.Role{Division: "Creative", Level: "staff"}}
	lead := permission.Actor{EmployeeID: "L", Role: permission.Role{Division: "Creative", Level: "lead"}}
	od := permission.Actor{EmployeeID: "O", Role: permission.Role{OD: true}}
	director := permission.Actor{EmployeeID: "D", Role: permission.Role{Director: true}}
	// Layered case: one employee is Creative Staff AND OD.
	staffOD := permission.Actor{EmployeeID: "SO", Role: permission.Role{Division: "Creative", Level: "staff", OD: true}}

	cases := []struct {
		name                           string
		a                              permission.Actor
		canWrite, admin, readAll, lead bool
		readOwnDiv, readOtherDiv       bool
	}{
		{"staff", staff, true, false, false, false, false, false},
		{"lead", lead, true, false, false, true, true, false},
		{"od", od, false, false, true, false, true, true},
		{"director", director, true, true, true, true, true, true},
		{"staff+od", staffOD, true, false, true, false, true, true},
	}
	for _, c := range cases {
		t.Run(c.name, func(t *testing.T) {
			if got := c.a.CanWrite(); got != c.canWrite {
				t.Errorf("CanWrite=%v want %v", got, c.canWrite)
			}
			if got := c.a.CanManageAdmin(); got != c.admin {
				t.Errorf("CanManageAdmin=%v want %v", got, c.admin)
			}
			if got := c.a.CanReadAll(); got != c.readAll {
				t.Errorf("CanReadAll=%v want %v", got, c.readAll)
			}
			if got := c.a.IsLead("Creative"); got != c.lead {
				t.Errorf("IsLead(Creative)=%v want %v", got, c.lead)
			}
			if got := c.a.CanReadDivision("Creative"); got != c.readOwnDiv {
				t.Errorf("CanReadDivision(Creative)=%v want %v", got, c.readOwnDiv)
			}
			if got := c.a.CanReadDivision("Ads"); got != c.readOtherDiv {
				t.Errorf("CanReadDivision(Ads)=%v want %v", got, c.readOtherDiv)
			}
		})
	}
}

// The layered OD scope must never confer write; only the underlying division
// scope does. A pure OD (no division) cannot write.
func TestLayeredOD_NeverWritesFromODScope(t *testing.T) {
	pureOD := permission.Actor{EmployeeID: "O", Role: permission.Role{OD: true}}
	if pureOD.CanWrite() {
		t.Fatal("pure OD must be read-only (CanWrite=false)")
	}
	staffOD := permission.Actor{EmployeeID: "SO", Role: permission.Role{Division: "Creative", Level: "staff", OD: true}}
	if !staffOD.CanWrite() {
		t.Fatal("Staff+OD must write from staff scope")
	}
	if staffOD.IsLead("Creative") {
		t.Fatal("Staff+OD is not a lead")
	}
}
